src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -571,6 +571,10 @@ namespace winrt::Microsoft::Terminal::Control::implementation
                 _searchBox->Open([weakThis = get_weak()]() {
                     if (const auto self = weakThis.get(); self && !self->_IsClosing())
                     {
+                        const auto displayInfo = DisplayInformation::GetForCurrentView();
+                        const auto scaleFactor = self->_core.FontSize().Height / displayInfo.RawPixelsPerViewPixel();
+                        const auto searchBoxRows = self->_searchBox->ActualHeight() / scaleFactor;
+                        self->_core.SetSearchScrollOffset(static_cast<int32_t>(std::ceil(searchBoxRows)));

```

Comment:
This won't refresh the row count if the DPI changes while the search box is open or the font size changes. We could just pass this value every time we `ControlCore::Search`. But since `ControlCore::Search` already has so many parameters, maybe we can add a `struct SearchRequest` and pass that instead?

What do you think? Does that improve the solution or does it make it worse?

Reply:
updated.  passing the offset with each call feels like a pretty big improvement (I don't need the field on ControlCore and don't have to deal with the _searchBox.open lambda and like you said minimizes the chances of a race condition),  just to double check the GetForCurrentView and RawPixelsPerViewPixel are pretty inexpensive? (my assumption is they are cached at the OS level)  The struct feels aesthetic,  I would be happy either way.

Reply:
> [...] the GetForCurrentView and RawPixelsPerViewPixel are pretty inexpensive? (my assumption is they are cached at the OS level)

Relative to other WinRT APIs? Oh yes, for sure.
Overall? lol, lmao even.

The `DisplayInformation` that gets returned by `GetForCurrentView()` is cached in the internal `StaticLifetimeStore`. That store is basically just an adapter access information on the `ApplicationModel.Core.CoreApplication` activation factory. So, it needs to get the activation factory, which is crazy expensive overall, but par for the course for WinRT APIs. There are at least 5 more `QueryInterface` calls and 2 "hashmap" (separate chaining 🙄) lookups in between this, to get the cached instance.

`RawPixelsPerViewPixel()` takes the `HWND` from that and calls `GetScaleFactorForWindow`, which is an internal shell API, and par for the course of shell, it gets an `IDeviceScalingInfo`, which does nothing but hook into your window message loop, wait for `WM_DPICHANGED` and cache that.

We can do that too, because we already have a message loop that receives `WM_DPICHANGED`.
Reading a `double` member from a class like `IslandWindow` or `IDeviceScalingInfo`: ~0.5ns.
`GetForCurrentView().RawPixelsPerViewPixel()`: ~6500ns.
WinRT and shell have outdone themselves yet again. Over 1000000% slower than it could be. Couldn't have done that if I tried.

I'm not sure whether it's worth it working around this, because WinUI and by extension WinRT are so slow overall, it'd be like trying to fix a leaky bucket. I'd be happy if we did it, because it's better than doing nothing, but the outlook is somewhat bleak for sure.

Reply:
thanks!  its a little bit over my head but I think I understand.  Are those all internal apis,  I tried to search through the WinUi3 repo but couldn't find anything.  Out of curiosity when do users start to notice stuff like that happening while they are typing,  65000ns still sounds find to me.

Reply:
There are many reasons why such a 6500ns are problematic:
* The API doesn't get called once. It gets called multiple times per second. (10x worse)
* This isn't an outlier: All the other WinRT APIs have similar overheads. (100x worse)
* ...and you call potentially thousands of such APIs per second to render your UI. (1000x worse)
* CPUs are all about thermal throttling nowadays. Intel's Thermal Velocity Boost for instance starts to throttle your CPU when it reaches 70°C.

Windows Terminal for instance, with no terminal tabs, and just the settings tab, burns somewhere around 1.5M cycles/s in the background. That's roughly 400'000ns of CPU. A bit more than 6500, right? When you move your cursor on top it, even if just on top of static elements, it burns about 800M cycles/s. That's roughly 200'000'000ns (200ms) of CPU time per second. That's a solid 20W of power draw just to figure out that nothing has changed.

I always find it curious that people always blame the CPUs for the poor battery life of Microsoft's laptops compared to MacBooks, but barely anyone realizes that it's because Windows just doesn't ever truly go idle. The constant boost velocity of the CPU that this causes is a battery killer.

Basically, it's not this one function that's a problem. It's [death by a thousand cuts](https://en.wikipedia.org/wiki/Creeping_normality).

Reply:
thank you!  that is really helpful learning wise for me.  You guys work on a totally different level than anything I have ever been exposed to.